### ISSUE TITLE
Add ability to filter the changes list

### ISSUE THREAD

**Issue by haacked** [desktop/desktop#6407]
**Please describe the problem you think should be solved**
TL;DR: Please add a way to filter the changes list so I can find the files I want to commit.

Hello friends! I'm a big fan of [The Thing Aboout Git](https://tomayko.com/blog/2008/the-thing-about-git). As such, I'll often go on manic sprees of making changes, maybe a few search and replaces throughout my code.

When I come up for air, I'm ready to make nice clean commits of logical groups of changes. The only problem is I have tons of changes, which is noisy. It makes it hard to find the changes I care about. I'd like a way to filter the changes.

**[Optional] Do you have any potential solutions in mind?**

Boy do I!

If you click on the History tab, you already have a nice text filter.

<img width="307" alt="screen shot 2018-12-12 at 2 43 03 pm" src="https://user-images.githubusercontent.com/19977/49903540-61506c00-fe1c-11e8-8989-c2a51665dc2e.png">
<img width="308" alt="screen shot 2018-12-12 at 2 43 21 pm" src="https://user-images.githubusercontent.com/19977/49903539-61506c00-fe1c-11e8-9aeb-dd03fdd29872.png">

It'd be nice to replicate that approach when you click on the Changes tab. Add a text box that filters the list.

You could get fancy with it and only show it when you have more than some number of changes (like 20).

For EXTRA CREDIT add the ability to filter on the type of change. For example, I sometimes want to filter out all deleted files when crafting a commit.

Thanks for your consideration. Big fan of the Desktop product!

**Comment by outofambit** [comment]
Hi @Haacked! Thanks so much for filing this issue. 💜 These seems useful to me (and mirrors a feature on dot com for pull requests). I'm going to mark this for design to take a look when they can. 😄 Thanks again.

**Comment by shiftkey** [comment]
@Kool-Koder also asked about this feature in #6941

**Comment by safield** [comment]
I think this is a really valuable feature! Anyone who works with large repos with large amounts of changed files, would consider this a feature a total must have!

**Comment by coltenkrauter** [comment]
I would love for this to be implemented. I would use it on the regular.

**Comment by leo2575** [comment]
For me just search based on userID would be a good feature

**Comment by steveward** [comment]
@syxcools requested this in https://github.com/desktop/desktop/issues/8578.

**Comment by haacked** [comment]
One thing that would be very helpful for this feature is a simple wildcard filter. For example, when I'm making a large commit, I often want to look at all the `*.csproj` files to make sure I didn't add anything to the wrong project. A filter on file extension works just as well.

**Comment by dperolio** [comment]
It's been almost 2 years now. Any news or updates? Anyone who's worked on any project of substantial size knows this is basically an essential feature... Wildcards and extras would be great, but even just the most basic text filtering would be a big help. Please add it.

**Comment by gingerbeardman** [comment]
I was just about to file a request for this. Thought I'd search for it first. No surprise that it's been sat on for years, that seems to be par for the course with this app and useful features suggested by the community. Oh well!

Scenario: 
I want to look at the changes in one file out of hundreds in my pending commit

Expected: 
be able to filter by filename

Actual: 
I have to scroll the list to find it

**Comment by Jaimi** [comment]
This would really be great when dealing with hundreds or thousands of changes -- many of which you don't currently want to commit. 
I work in Unity a lot, and it can randomly decide to update assets. Sometimes you want to keep these updates, and sometimes not. Being able to temporarily filter to just source code would be a great addition. It's extremely painful to dig through a thousand changes (I kid you not) to find the two files you need to check in.

**Comment by malan-x** [comment]
For two years, the developers have not been able to add the simplest filtering function. A feature that would save users a lot of time. Are you sadistic? Those you can add filtering functionality to the branches section. Although there are fewer branches in most projects than fingers on a hand. But you cannot add search by commits or by changes, which are typed in thousands. Do you use your own product at all?

**Comment by gingerbeardman** [comment]
I think they do use it, but are only interested to add features that support their own workflows (where they probably have multiple branches per dev and multiple devs).

**Comment by teanfoo** [comment]
@steveward Help me.

**Comment by billygriffin** [comment]
Thanks everyone for the feedback here! We discussed this today and I wanted to share some thoughts and ask a question about how you'd expect this to work. One consideration we're trying to think through is what happens when you have a filter applied. You see the changes that meet the filter criteria, but then the rest of the files will not be shown, so it's tricky to figure out what will actually be committed if you're committing in a filtered state. 

For the filter itself, based on this issue, it seems like it's worth supporting both file name and extension filtering. So you can type `.md` and all markdown files will show up, or you can type `README` and your README.md file will show up (but only files that actually have changes, obviously).

We've got a few options for how we handle committing when a filter is applied:

### Anything goes (commit everything in and out of the filter)

The benefit of this is that you're in control over your selection. But I think the downside risk is too great that you aren't able to see what files will actually be committed if for example, there are some files selected, then you filter, then select a few more, then commit. It makes it too unclear what's actually going to be committed and that feels like too many dragons.

### Don't allow commits when the filter is applied

This would have the benefit of making it more explicit what's being committed and that the filter is only for finding and selecting things. But the downside is convenience - you'd have to cancel the filter in order to commit.

### From a filtered state, only commit files that show up in your filter

This is an attempt to mesh the two. Only the changes shown would be committed, and changes from files not included in the filter would not be included in the commit. This has a similar downside as the first option except in the opposite direction. If you _think_ it's going to commit everything, including what's in and out of the filter, then you'll be surprised that it only committed the changes that were part of the filter.

If you have other ideas about how this might work, we'd love to hear them. Regardless of the approach we take, it'll be important to communicate the expectation so people understand what will happen if they try to commit a filtered list of changes.

**Comment by Jaimi** [comment]
Personally, I think the correct process is the last one -- only commit files that are visible. I would go further and say to make it easier you could just clear the selection when applying the filter. 
The purpose of this, at least for me, is to filter out all the chaff, so I can see what I want to commit. I could not care less about the unfiltered data. I would never want to commit those items. 
Here is my preferred workflow:

Enter *.cs in the filter box to display only code changes. Select the ones I want (probably all of them), and hit commit, committing just those items.

**Comment by gingerbeardman** [comment]
Agreed. To filter is to focus on those files, unfiltered files are to be ignored during filter.

**Comment by dperolio** [comment]
While not exactly _ideal_, the "Don't allow commits when the filter is applied" option definitely makes the most sense, at least for my use cases, and it's the solution that would cause the least confusion and be the least user error prone, I would think. I personally don't think there is a particular time where I would want to only commit a set of filtered items, or if so, very rarely. Even in the event that was the case though, you could still do this with the "Don't allow commits when the filter is applied" option, you would just have to [clear any currently selected items] enter your filter, select all the items (preferably with a simple checkbox as seen in my mockup below), clear the filter (by clicking the x in the filter box), and commit.

For my case, when I have hundred(s) (or even thousands) of changed files, and I may be looking for specific files to commit. Let's say I have 500 changed files, and I've already selected 50 files I want to commit, but can't find another 20 or so I'm looking for; I would just type a filter, maybe something as simple as pie.js, select the file, then change the filter, repeat until I have all the files I'm looking for selected, then commit the changes. So it would be just like it is now, but make finding certain files quicker/easier, and that should be the goal, to me.

One possible UI for such a thing would be to have the "main" section collapse (or "appear" to be collapsed) to the bottom when the user enters anything in the filter search bar, and a new section header would appear under the filter search bar with how many filtered files are found, and a checkbox to deselect/select all found filtered items. I also added a "x files selected" to the main category header to be able to easily tell how many files are currently selected to be committed which should help clarify things to the user and avoid confusion.

![GitHub Desktop Filter Mockup](https://user-images.githubusercontent.com/3791055/119598031-29e99600-bdb0-11eb-932f-4d7aa1077a8f.png)

With this setup, you could potentially even have the "Anything goes (commit everything in and out of the filter)" work, if you have the section headers here be like collapsible accordions--- i.e. clicking the "544 changed files, 56 files selected" header could expand to right below the filter header, collapsing the filtered section. I feel like just having the "x files selected" info in the main section header should make it much more clear to the user what they're committing, at least in a ballpark sense.

![GitHub Desktop Filter Accordion](https://user-images.githubusercontent.com/3791055/119599344-d0369b00-bdb2-11eb-8d11-c79ad3f88e8d.png)

**Comment by haacked** [comment]
When using Desktop, I think of the GUI as being similar to staging a commit. So to create a commit, I might apply one filter, select some files or changes in files, then apply another filter, and select more changes. When I commit, I'd expect everything I selected to be committed, whether I was currently in a filter or not. To me, a filter is a just a filtered _view_ of my changes and shouldn't change the underlying state of the commit that I'm crafting. In the same way the changes that aren't visible because you scrolled down (when there are too many commits to fit on the screen) are still included in the commit.  So I'm in favor of option one or two.

Option one could work if the text above the commit box that summarizes the number of changes being committed also included the number of changes hidden by the current filter that will be included. That would let the user know they're not seeing the full commit because of a filter.

Option two could be less onerous if the commit button becomes a "Clear filter" button. Click it, then click it again to commit. No modals.

Since filters help you craft a commit, it may make sense to clear the filter when you create the commit no matter which option you end up implementing.

**Comment by Jaimi** [comment]
If you plan to have invisible selected items, it's imperative that the user knows about them when committing. Already the application will preselect changes for you. If you forget to clear them before filtering, you will end up committing a bunch of changes you don't expect to. 
Maybe it's best if the filter shows all selected items + whatever the filter is. Or you have a dialog when committing that says "You have N unshown items selected that will be committed" 
But personally, I think committing invisible items is a _bad idea_.

**Comment by haacked** [comment]
@billygriffin another option to consider. A filter on the changes list never hides changes. Instead, it sorts the changes - first those that match the filter followed by those that don't - and applies different styles to those that match and those that don't match the filter.

Render the changes that don't match in a less prominent style (reduced opacity, greyed out, etc) so it's obvious that they don't match the filter. Obviously the changes that match the filter should be more prominent to make it obvious they do match.

This addresses the use case I care about which is being able to find the changes I'm concerned about as they will rise to the top. But it also doesn't change how committing works. You can still select the filtered out changes for inclusion in the commit.

**Comment by gingerbeardman** [comment]
Still need this daily.

**Comment by jozydapozy** [comment]
Really need this feature. When refactoring large projects it's really a must have!

> Don't allow commits when the filter is applied

This makes the most sense I think, because it would be dangerous to commit selected files that are not visible. That's not an obvious action for a user unkown of the functionality.

Maybe even a better approach: 
> Don't allow commits when the filter is applied **AND 1 or more invisible files are selected**, otherwise, allow it.

**Comment by tidy-dev** [comment]
From #15036,
### The feature request

Hello, will be great to have a select menu (or some similar) to filter by type. 
I need to see only removed file... updated or added.

Many thanks for your hard work :)

Dario

![Screenshot 2022-07-26 105721](https://user-images.githubusercontent.com/17809213/181065958-c32d49f0-e28c-48ed-833a-891f8032e1ce.png)


### Proposed solution

Have a select menu or radio button to show/hide toggle by type at top UI

**Comment by Landeplage** [comment]
Was going to request a filter changes feature, then found this thread. Would be super useful! Are there any plans to add this?

**Comment by DanForever** [comment]
I'd like to also throw my input into the ring if that's ok.

I use Perforce (Helix? P4? it seems to have a few names) in my day job, and that's designed around the concept of "changelists", which is basically a way of grouping changes for ease of management and submission.

The concept of "changelists" runs through the core of Perforce and inhabits every action you do. Git of course is a very different beast. But there's no reason why we can't take the concept of grouped changes and simply apply it to a gui interface (like Github desktop).

The way it would work is that the changes tab would be a tree. To start with you'd have one root node called "Default" into which all new changes automatically fall.

At any time, a user can create a new group (which will become a new root node), and changes (files) can be dragged and dropped between the groups.

Groups can have their own summaries and descriptions, which means you can fill those out when the group is created (if that's more convenient to the user) and simply submit when the changes are ready to go.

Users can submit a whole group, or individual files. If a group is submitted, just remove it from the list as changes do currently, but if individual files are submitted, then you need to decide what to do with the remaining files in the group.
You could either send them back to the default changelist and delete the group they were in, or simply keep the group around.
My preference would be for keeping the group, but perhaps erasing the summary and description unless the user explicitly chooses not to.

**Comment by tidy-dev** [comment]
From #15403,


> if i have a long list of changed files (10-100s), there isn't the ability to jump to a specific file in that list that I could find. I have to select all the files, right click, copy paths to a text editor, search there, and then try to correlate the line number to the actual list back in my GH desktop (OSX).

**Comment by KingJavo** [comment]
I'd really love to have a way to filter by extension in the history/changes tab as well. Mine is flooded with thousands of project .yy file changes from gamemaker studio because they're IDE has an issue where it shows changes just for opening your project that are not really substantial changes worth noting.

PLEASE, PLEASE, PLEASE can you add this as a feature?

**Comment by dustin-rcg** [comment]
+1 Looking for this again today. I have 158 changed files, and I can't even sort them alphabetically (by file name not path), let alone search them.

**Comment by EmirLogas** [comment]
It would be very good to add the search feature in all changes.

**Comment by marz-uw** [comment]
Man, we're almost **at half-a-decade since this this request was opened** (and the pain was probably around prior to this), with no progress

This feels like such a willful blindness to a core pain point for anyone who uses the GitHub Desktop UI

**Comment by Ozcreates** [comment]
This functionality is a must, I dont understand why it doesn't exist in the first place

**Comment by yaakov1111** [comment]
why the BIG delay on this feature...

**Comment by solsnare** [comment]
Switching off this because of this BS.  Why are there so many core features of a git client missing.  Why can't I have multiple stashes, why can't I sort through changes, why TF does it merge non text files automatically and corrupt everything in Unity.  What is wrong with this tool.

**Comment by tidy-dev** [comment]
Requested in https://github.com/desktop/desktop/issues/17774

**Comment by Jaimi** [comment]
FIVE YEARS LATER.  Still no filter. How hard can this be?

When displaying changes:

1. Is it in the list?  Yes: Display it. No: Don't display it, and unselect it.

Sorry for numbering, it turned out to be only one step.  How much code is that?  If I were a web developer, I would do it myself.

**Comment by 0xdevalias** [comment]
RE: https://github.com/desktop/desktop/issues/6407#issuecomment-847360540

> If you have other ideas about how this might work, we'd love to hear them. Regardless of the approach we take, it'll be important to communicate the expectation so people understand what will happen if they try to commit a filtered list of changes.

Personally, I would expect the 'Anything goes (commit everything in and out of the filter)' option; BUT, with caveats:

- the UI should definitely be showing some indication of what I have selected, even if the filter is hiding them
  - This could be something like https://github.com/desktop/desktop/issues/6407#issuecomment-848426014
  - Or it could be as simple as `XX changed files, YY files selected (of which ZZ are hidden by current filter)`
- when I click commit, an alert pops up re-confirming that I know that `ZZ` files hidden by the current filter will be committed, and asks me to confirm
  - This should have an option to "never show me this again"

That way you're not hiding anything/doing anything overly unexpected, there is a warning to ensure newer/unfamiliar users aren't caught off guard by it, and more experienced users can 'reduce the noise' by disabling the warning because they know what they're doing.

**Comment by steveward** [comment]
Requested in https://github.com/desktop/desktop/issues/18299.

**Comment by spartan-jack** [comment]
How about we get the most basic of functionality and let the feedback shape its design instead of waiting over half a decade for an extremely basic feature?

**Comment by Jaimi** [comment]
Six years now. Six years of hunting through hundreds of unwanted changes (thanks, Unity), without even the ability to sort here. Everyone says "just put everything else on your ignore list, you're an idiot", but the fact is that sometimes I do want and need to commit those items. Just not all the time.   You don't have to make something that works for everyone all the time and covers 100% of use cases.  The simplest - filtering clears the selection - would do 99% of what everyone needs. Even those that need other items commited, etc - because they could commit one, re-filter and commit again, then push everything at once.
Heck, even sorting by extension would be a major improvement.

**Comment by solsnare** [comment]
> Six years now. Six years of hunting through hundreds of unwanted changes (thanks, Unity), without even the ability to sort here. Everyone says "just put everything else on your ignore list, you're an idiot", but the fact is that sometimes I do want and need to commit those items. Just not all the time. You don't have to make something that works for everyone all the time and covers 100% of use cases. The simplest - filtering clears the selection - would do 99% of what everyone needs. Even those that need other items commited, etc - because they could commit one, re-filter and commit again, then push everything at once. Heck, even sorting by extension would be a major improvement.

Highly suggest ya'll just go use Fork, can evaluate it indefinitely and its the best one i've found so far.  

Also how tf do i stop getting emails on these replies lol

**Comment by Jaimi** [comment]
> > Six years now. Six years of hunting through hundreds of unwanted changes (thanks, Unity), without even the ability to sort here. Everyone says "just put everything else on your ignore list, you're an idiot", but the fact is that sometimes I do want and need to commit those items. Just not all the time. You don't have to make something that works for everyone all the time and covers 100% of use cases. The simplest - filtering clears the selection - would do 99% of what everyone needs. Even those that need other items commited, etc - because they could commit one, re-filter and commit again, then push everything at once. Heck, even sorting by extension would be a major improvement.
> 
> Highly suggest ya'll just go use Fork, can evaluate it indefinitely and its the best one i've found so far.
> 
> Also how tf do i stop getting emails on these replies lol

Unfortunately, not even Fork can filter the pending changes. Or at least, that's what Fork Support told me when I queried them. Also, if you're on the web (follow the link) on the top right of this issue, hit the "Unsubscribe" button.

**Comment by solsnare** [comment]
> > > Six years now. Six years of hunting through hundreds of unwanted changes (thanks, Unity), without even the ability to sort here. Everyone says "just put everything else on your ignore list, you're an idiot", but the fact is that sometimes I do want and need to commit those items. Just not all the time. You don't have to make something that works for everyone all the time and covers 100% of use cases. The simplest - filtering clears the selection - would do 99% of what everyone needs. Even those that need other items commited, etc - because they could commit one, re-filter and commit again, then push everything at once. Heck, even sorting by extension would be a major improvement.
> > 
> > 
> > Highly suggest ya'll just go use Fork, can evaluate it indefinitely and its the best one i've found so far.
> > Also how tf do i stop getting emails on these replies lol
> 
> Unfortunately, not even Fork can filter the pending changes. Or at least, that's what Fork Support told me when I queried them. Also, if you're on the web (follow the link) on the top right of this issue, hit the "Unsubscribe" button.

I've been using it for a while now and it indeed can search through changes and even allows u to search through commits which is amazing

**Comment by Stexe** [comment]
@sergiou87 : Is there a reason why this is not planned? I'm running into this problem myself. It will be like "this 1 file is preventing you from pulling so you need to stash it" -- I just want to delete the changes I made to said file (instead of stashing), but now I have to sort through 300+ changes to find that single file to remove changes.

Being able to filter changed files by name would save SOOOOOOOOOOOOO much time. I'm kinda flabbergasted that this change hasn't been implemented in 5+ years of people asking for it. Heck, even being able to sort changes by different filters would be a step in the right direction...

**Comment by sergiou87** [comment]
It just means it's not planned, not that we won't ever work on it. We are a very small team and have other priorities right now. At some point we will have time for other stuff, like this. Maybe in the next few months? Who knows 😅

**Comment by gingerbeardman** [comment]
Obviously you have other priorities right now, but also for the entirety of the last 6 years? Let's not kid ourselves that this has any remote likelihood of being worked on in the next few months.

**Comment by Ozcreates** [comment]
I don't know if it's very hard to add since you're a small team and I appreciate all the work but I must add that this should definitely be a priority, because that's literally the only thing missing for me and all the people I know using github desktop

**Comment by ghnp5** [comment]
wow... some people are really entitled here... you must be paying loads of money to the developers! 🙄

Thank you, devs, for GitHub Desktop. It's been a really useful tool to me, together with VSCode. <3

---

Anyway - I always had the impression this change was to implement a Search/Filter to the Commit History, but looks like the intention here is to filter the uncommitted changes?

I've just done a quick search on issues, and seems #7022 is the one I was looking for, and that would be much more useful to me :)  just subscribed that one!

**Comment by steveward** [comment]
Feedback gathered from https://github.com/desktop/desktop/issues/19075:

> The feature request
> I would love to be able to "hide" unwanted files from the Changes tab. I don't want them fully ignored and added to the .gitignore, and I can't commit those files either because that would mess up other contributor's work. I just want to temporarily add them to a hidden list so my local copy of Github Desktop will always mark them as excluded and hidden from the Changes tab until the next time Github Desktop is opened, or hidden files is toggled off.
> 
> Proposed solution
> My changes tab often gets cluttered with files that Unity generates automatically and it gets annoying to constantly work around those files, making sure they are unchecked and will not be committed. It's nice that Github desktop already doesn't recheck those files if I've made further changes, but if I made a couple changes at once and I want to make separate commits for different files, it's frustrating to wade through all the files I know that I won't ever commit.
> 
> Additional context
> I work on a large project with many team members in Unity. Often after pulling from main, Unity will generate some atlas or meta files that I do not want to recommit and they will constantly appear in the Changes tab and I have to go through and uncheck them all manually. The tech lead said it would be annoying to fix just for me, so having this option on my end would be awesome.

and


> I think filtering would work because usually I'm working in specific folders so, perhaps if I could filter out everything but the folders i'm working in would be great.
> 
> Then for it to be truly useful, I can save filters and set one as default so it just auto-filters out as I'm working.

**Comment by gingerbeardman** [comment]
Welcome to the thread, enjoy your stay

**Comment by tidy-dev** [comment]
Another potential way to filter changes list:  [Not see files that are only whitespace changes.](https://github.com/desktop/desktop/issues/15980)

**Comment by tidy-dev** [comment]
From #20163 

>The feature request
>In Unity, each file has a .meta file accompanying it. This can clutter the Github Desktop changelist and duplicate information.

>I think it'd be a good idea to add a way to hide them by default, while keeping them open for inspection on demand.

>Proposed solution
>There could be a "Hidden patterns" panel (maybe marking patterns as hidden-on-gui=true in .gitattributes):

>Files and folders that match these patterns will still be pushed, but will not clutter the changelist.
>On the GUI, they will appear by default, and the user can click a checkbox to hide them.


>After the user confirms hiding "hidden" files, the hidden files will not appear there.

>When the checkbox is checked, ALL files should appear as normal.
>When unchecked, only non-hidden files should appear.
>Github reviewers could also opt to hide hidden files in one go, for a cleaner and painless workflow.
>*The current workflow for reviews involves clicking the "filter items" button and unchecking .meta files.

**Comment by samimsu** [comment]
I cloned the repo and found that it's added as a beta feature. I assume you can get it from https://desktop.github.com/beta/

<img src="https://github.com/user-attachments/assets/f8792aac-dd2b-4968-9110-3acd13e7a19c" width="200" alt="" />

**Comment by gingerbeardman** [comment]
Can confirm it's in version 3.4.19-beta1 (arm64) downloaded from https://desktop.github.com/beta/

**Comment by Jaimi** [comment]
Yes! It's here, in the beta channel! And it works great - thanks!

**Comment by haacked** [comment]
Closed by https://github.com/desktop/desktop/pull/20220. Just tried it out. It's great!

**Comment by CecilePalant** [comment]
Hi, sorry to express opposing opinion here (I might not be alone, just one to vocalise). I have no use for this and find that it adds noise to a previously very sleek UI. I understand that some may find it super useful: could you please make it as a toggle option in the settings?

**Comment by gingerbeardman** [comment]
@CecilePalant see #20303
